metabase/metabase
# fix(sdk): fix ResizeObserver error in the ExplicitSize component

Closes https://github.com/metabase/metabase/issues/43304

### Description

In the Cypress CI and in the Embedding SDK, we often see this error: "ResizeObserver loop completed with undelivered notifications."

This is caused by layout thrashing in our use of resizeObserver in the ExplicitSize component. Per https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver#observation_errors

> As long as the error event does not fire indefinitely, resize observer will settle and produce a stable, likely correct, layout. However, visitors may see a flash of broken layout, as a sequence of changes expected to happen in a single frame is instead happening over multiple frames. If you want to prevent these errors, the solution depends on what your intended effect is. If you actually intend to have an infinite loop, you just need to defer the resizing code in your ResizeObserver callback to after the browser repaints. You can put it into a requestAnimationFrame callback.

### Demo

Before

![CleanShot 2567-09-04 at 16 16 48@2x](https://github.com/user-attachments/assets/1fc97ba3-23a5-4d5e-be64-611559d086a5)

After

![CleanShot 2567-09-04 at 16 30 38@2x](https://github.com/user-attachments/assets/57b1a7e9-5109-4c37-bff5-d5ed07a793f4)

### How to verify

- Run the SDK with create-react-app. Don't use Vite, or you won't see the error.
   - Option 1: You can test this with Jeff's example here: https://github.com/metabase/metabase-nodejs-react-sdk-embedding-sample.
   - Option 2: you can run `npx @metabase/embedding-sdk-react@latest start` in your own create-react-app setup. IMO this is a bit faster.
- You should see the error as per the above screenshot without this PR.
- With this PR, the above error should go away.
   - PS. If someone has ideas on how to regression-test this let me know. I can't even reproduce it on Cypress CI, so no idea how to test against that.
